gmso/atomtyping/parameterize.py

```diff
@@ -0,0 +1,284 @@
+        sub_top_label = next(
+            iter(
+                nx.classes.function.get_node_attributes(
+                    top_graph, "atom_data"
+                ).values()
+            )
+        ).subtopology_label
+        ff_of_interest = forcefields[sub_top_label]
+        at_rules_provider = get_atomtyping_rules_provider(
+            gmso_ff=ff_of_interest
+        )
+        sub_type_map = find_atomtypes(top_graph, at_rules_provider)
+        traverse_typemap(top, sub_type_map, ff_of_interest)
+        for graph, mapping in identical:
+            for node in graph:
+                mapped = sub_type_map[mapping[node]]["atom_type"]
+                top._sites[node].atom_type = mapped.clone()
+
+
+def traverse_typemap(top, type_map, forcefield):
+    """Take a typemap and applies those atomtypes to the Topology.
+
+    Parameters
+    ----------
+    top: gmso.core.topology.Topology, required
+        The GMSO topology on which to apply atomtypes
+    typemap: dict, required
+        The dictionary of atom_index with iterated matching to get the atomtype of that site
+    forcefield: gmso.forcefield.Forcefield, required
+        The GMSO forcefield object which has information associated with each unique atomtype
+    """
+    for index in type_map:
+        site_of_interest = top._sites[index]
+        site_of_interest.atom_type = forcefield.atom_types[
+            type_map[index]["atomtype"]
+        ].clone()
+        type_map[index]["atom_type"] = site_of_interest.atom_type
+
+
+def apply_residue_names(top, more):  # TODO
+    """Take a topology and applies residue info to all sites.
+
+    Parameters
+    ----------
+    top: gmso.core.topology.Topology, required
+        The GMSO topology on which to apply atomtypes
+    more: TBD, required
+        The information necessary to map residue name and number for each site
+    """
+    return
+
+
+def apply_connection_types(top, forcefields, parameter):
+    """Set Topology connection types from matching atomclass information in forcefield.
+
+    Parameters
+    ----------
+    top: gmso.core.topology.Topology, required
+        The GMSO topology on which to apply paramter_types
+    forcefields: dict, required
+        The keys are labels that match the subtopology labels for each molecule, and the
+        values are gmso Forcefield objects that gets applied to the specified molecule
+    parameter: str, required
+        The connection type to parameterize. Can any of "bond", "angle", "dihedral", "improper"
+    """
+    connect_members = {
+        "bond": [0, 1],
+        "angle": [0, 1, 2],
+        "dihedral": [0, 1, 2, 3],
+        "improper": [0, 1, 2, 3],
+    }  # TODO validate improper order
+    molecule_label = "label_"  # place to grab info referencing molecule name
+    for connect in getattr(top, "_" + parameter + "s"):
+        sub_top_label = connect.connection_members[0].get(molecule_label)
+        if sub_top_label is None:
```

Comment:
```suggestion
        if not sub_top_label:
```
I think this would also catch the case where the `sub_top_label` is empty string.